File: fullstack-fintech-app/resiliences-service/Controllers/BeneficiaryController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using resiliences_service.Enums;
using resiliences_service.interfaces;
using resiliences_service.Models;
using resiliences_service.Payloads;

namespace resiliences_service.Controllers
{
    [ApiController]
    [Route("beneficiaries")]
    [Authorize]
    public class BeneficiaryController : ControllerBase
    {
        private readonly IBeneficiaryService            _service;
        private readonly ILogger<BeneficiaryController> _logger;

        public BeneficiaryController(IBeneficiaryService service, ILogger<BeneficiaryController> logger)
        {
            _service = service;
            _logger  = logger;
        }

        // POST /beneficiary/create
        [HttpPost("create")]
        public async Task<IActionResult> Create([FromBody] CreateBeneficiaryRequest req)
        {
            if (!ModelState.IsValid)
                return BadRequest(new { message =  "Invalid request payload" });

            if (req.BeneficiaryType == BeneficiaryType.bank)
            {
                if (string.IsNullOrEmpty(req.AccountNumber))
                    return BadRequest(new { message =  "accountNumber is required for bank beneficiaries" });
                if (string.IsNullOrEmpty(req.AccountName))
                    return BadRequest(new { message =  "accountName is required for bank beneficiaries" });
                if (string.IsNullOrEmpty(req.BankCode))
                    return BadRequest(new { message =  "bankCode is required for bank beneficiaries" });
                if (string.IsNullOrEmpty(req.BankName))
                    return BadRequest(new { message =  "bankName is required for bank beneficiaries" });
            }
            else if (req.BeneficiaryType == BeneficiaryType.user)
            {
                if (string.IsNullOrEmpty(req.RecipientUsername))
                    return BadRequest(new { message =  "recipientUsername is required for user beneficiaries" });
            }

            var beneficiary = new Beneficiary
            {
                UserId          = req.UserId,
                BeneficiaryType = req.BeneficiaryType,
                BeneficiaryName = req.BeneficiaryName,
                Currency        = req.Currency,
                IsActive        = true
            };

            if (req.BeneficiaryType == BeneficiaryType.bank)
            {
                beneficiary.AccountNumber     = req.AccountNumber;
                beneficiary.AccountName       = req.AccountName;
                beneficiary.BankCode          = req.BankCode;
                beneficiary.BankName          = req.BankName;
                beneficiary.RecipientUsername = null;
            }
            else
            {
                beneficiary.RecipientUsername = req.RecipientUsername;
                beneficiary.AccountNumber     = null;
                beneficiary.AccountName       = null;
                beneficiary.BankCode          = null;
                beneficiary.BankName          = null;
            }

            try
            {
                await _service.CreateAsync(beneficiary);
                return StatusCode(201, new { status = "success", message = "Beneficiary created successfully" });
            }
            catch (InvalidOperationException ex)
            {
                return BadRequest(new { message =  ex.Message });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "[BeneficiaryController] Create failed");
                return StatusCode(500, new { message =  "Failed to create beneficiary" });
            }
        }

        // GET /beneficiaries/{id}
        [HttpGet("{id}")]
        public async Task<IActionResult> GetById(uint id)
        {
            try
            {
                var beneficiary = await _service.GetByUserIdAsync(id);
                if (beneficiary == null)
                    return NotFound(new { message =  "Beneficiary not found" });

                return Ok(new { status = "success", message = "Beneficiary retrieved successfully", data = beneficiary });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "[BeneficiaryController] GetById failed");
                return StatusCode(500, new { message =  "Failed to fetch beneficiary" });
            }
        }

        // GET /beneficiaries/{userId}/all
        [HttpGet("{userId}/all")]
        public async Task<IActionResult> GetAllByUserId(uint userId)
        {
            try
            {
                var beneficiaries = await _service.GetAllByUserIdAsync(userId);
                return Ok(new { status = "success", message = "Beneficiaries retrieved successfully", data = beneficiaries });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "[BeneficiaryController] GetAllByUserId failed");
                return StatusCode(500, new { message =  "Failed to fetch beneficiaries" });
            }
        }

        // GET /beneficiaries/{userId}/type?type=bank
        [HttpGet("{userId}/type")]
        public async Task<IActionResult> GetByType(uint userId, [FromQuery] string type)
        {
            if (!Enum.TryParse<BeneficiaryType>(type, true, out var beneficiaryType) ||
                (beneficiaryType != BeneficiaryType.bank && beneficiaryType != BeneficiaryType.user))
                return BadRequest(new { message =  "Type must be 'bank' or 'user'" });

            try
            {
                var beneficiaries = await _service.GetByTypeAsync(userId, beneficiaryType);
                return Ok(new { status = "success", message = "Beneficiaries retrieved successfully", data = beneficiaries });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "[BeneficiaryController] GetByType failed");
                return StatusCode(500, new { message =  "Failed to fetch beneficiaries" });
            }
        }

        // GET /beneficiaries/{userId}/search?search=query
        [HttpGet("{userId}/search")]
        public async Task<IActionResult> Search(uint userId, [FromQuery] string search)
        {
            if (string.IsNullOrEmpty(search))
                return BadRequest(new { message =  "Provide a search query parameter" });

            try
            {
                var beneficiaries = await _service.SearchAsync(userId, search);
                return Ok(new { status = "success", message = "Search completed successfully", data = beneficiaries });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "[BeneficiaryController] Search failed");
                return StatusCode(500, new { message =  "Failed to search beneficiaries" });
            }
        }

        // GET /beneficiaries/{id}/verify
        [HttpGet("{id}/verify")]
        public async Task<IActionResult> Verify(uint id)
        {
            try
            {
                var beneficiary = await _service.GetByUserIdAsync(id);
                if (beneficiary == null)
                    return NotFound(new { message =  "Beneficiary not found" });

                return Ok(new { status = "success", message = "Beneficiary verified", data = beneficiary });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "[BeneficiaryController] Verify failed");
                return StatusCode(500, new { message =  "Failed to verify beneficiary" });
            }
        }

        // PUT /beneficiaries/{id}
        [HttpPut("{id}")]
        public async Task<IActionResult> Update(uint id, [FromBody] CreateBeneficiaryRequest req)
        {
            if (!ModelState.IsValid)
                return BadRequest(new { message =  "Invalid request payload" });

            var beneficiary = new Beneficiary
            {
                Id              = id,
                UserId          = req.UserId,
                BeneficiaryType = req.BeneficiaryType,
                BeneficiaryName = req.BeneficiaryName,
                Currency        = req.Currency
            };

            if (req.BeneficiaryType == BeneficiaryType.bank)
            {
                beneficiary.AccountNumber     = req.AccountNumber;
                beneficiary.AccountName       = req.AccountName;
                beneficiary.BankCode          = req.BankCode;
                beneficiary.BankName          = req.BankName;
                beneficiary.RecipientUsername = null;
            }
            else
            {
                beneficiary.RecipientUsername = req.RecipientUsername;
                beneficiary.AccountNumber     = null;
                beneficiary.AccountName       = null;
                beneficiary.BankCode          = null;
                beneficiary.BankName          = null;
            }

            try
            {
                await _service.UpdateAsync(beneficiary);
                return Ok(new { status = "success", message = "Beneficiary updated successfully" });
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(new { message =  ex.Message });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "[BeneficiaryController] Update failed");
                return StatusCode(500, new { message =  "Failed to update beneficiary" });
            }
        }

        // DELETE /beneficiaries/{id}
        [HttpDelete("{id}")]
        public async Task<IActionResult> Delete(uint id, [FromQuery] uint userId)
        {
            try
            {
                await _service.DeleteAsync(id, userId);
                return Ok(new { status = "success", message = "Beneficiary deleted successfully" });
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(new { message =  ex.Message });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "[BeneficiaryController] Delete failed");
                return StatusCode(500, new { message =  "Failed to delete beneficiary" });
            }
        }

        // DELETE /beneficiaries/bulk
        [HttpDelete("bulk")]
        public async Task<IActionResult> DeleteByIds([FromBody] DeleteBeneficiariesRequest req)
        {
            if (req.Ids == null || req.Ids.Count == 0)
                return BadRequest(new { message =  "Provide a valid list of IDs" });

            try
            {
                await _service.DeleteByIdsAsync(req.Ids, req.UserId);
                return Ok(new { status = "success", message = "Beneficiaries deleted successfully" });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "[BeneficiaryController] DeleteByIds failed");
                return StatusCode(500, new { message =  "Failed to delete beneficiaries" });
            }
        }

        // GET /beneficiaries/{userId}/username/{recipientUsername}
        [HttpGet("{userId}/username/{recipientUsername}")]
        public async Task<IActionResult> CheckUserBeneficiary(uint userId, string recipientUsername)
        {
            if (string.IsNullOrWhiteSpace(recipientUsername))
                return BadRequest(new { message =  "recipientUsername is required" });

            try
            {
                var beneficiary = await _service.GetByUserIdAndUsernameAsync(userId, recipientUsername);
                
                if (beneficiary == null)
                    return NotFound(new { message =  "Beneficiary not found" });

                return Ok(new
                {
                    status = "success",
                    message = "Beneficiary retrieved successfully",
                    data = beneficiary
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "[BeneficiaryController] CheckUserBeneficiary failed");
                return StatusCode(500, new { message =  "Failed to check beneficiary" });
            }
        }

    }
}

File: epay/beneficiary/src/main/java/com/epay/beneficiary/controller/BeneficiaryController.java
package com.epay.beneficiary.controller;

import com.epay.beneficiary.service.BeneficiaryService;
import com.epay.common.exception.ApiResponse;
import com.epay.common.exception.BadRequestException;
import com.epay.common.exception.ErrorCode;
import com.epay.domain.beneficiary.dto.BeneficiaryDTO;
import com.epay.domain.beneficiary.enums.BeneficiaryType;
import com.epay.domain.beneficiary.input.CreateBeneficiaryRequest;
import com.epay.domain.beneficiary.input.DeleteBeneficiariesRequest;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Tag(name = "Beneficiaries", description = "Save and manage transfer beneficiaries (internal users and external banks)")
@Slf4j
@RestController
@RequestMapping("/beneficiaries")
@RequiredArgsConstructor
public class BeneficiaryController {

    private final BeneficiaryService beneficiaryService;

    @Operation(
        summary     = "Save a new beneficiary",
        description = "Creates a beneficiary record. For BANK type, accountNumber, accountName, bankCode and bankName are required. For USER type, recipientUsername is required."
    )
    @PostMapping("/create")
    @PreAuthorize("hasRole('USER')")
    public ResponseEntity<ApiResponse<BeneficiaryDTO>> create(
            @Valid @RequestBody CreateBeneficiaryRequest req) {
        validateCreateRequest(req);
        BeneficiaryDTO dto = beneficiaryService.create(req);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.success("Beneficiary created successfully", dto));
    }

    @Operation(
        summary     = "List all beneficiaries for a user",
        description = "Returns every saved beneficiary (BANK and USER types) for the given user ID."
    )
    @GetMapping("/{userId}/all")
    @PreAuthorize("hasRole('USER')")
    public ResponseEntity<ApiResponse<List<BeneficiaryDTO>>> getAllByUserId(
            @PathVariable Long userId) {
        return ResponseEntity.ok(ApiResponse.success("Beneficiaries retrieved successfully",
                beneficiaryService.getAllByUserId(userId)));
    }

    @Operation(
        summary     = "Get a beneficiary by ID",
        description = "Returns a single beneficiary record for the given ID, scoped to the requesting user."
    )
    @GetMapping("/{id}")
    @PreAuthorize("hasRole('USER')")
    public ResponseEntity<ApiResponse<BeneficiaryDTO>> getById(
            @PathVariable Long id,
            @RequestParam Long userId) {
        return ResponseEntity.ok(ApiResponse.success("Beneficiary retrieved successfully",
                beneficiaryService.getById(id, userId)));
    }

    @Operation(
        summary     = "List beneficiaries by type",
        description = "Filters the user's beneficiaries by type. Valid values: BANK or USER."
    )
    @GetMapping("/{userId}/type")
    @PreAuthorize("hasRole('USER')")
    public ResponseEntity<ApiResponse<List<BeneficiaryDTO>>> getByType(
            @PathVariable Long userId,
            @RequestParam String type) {
        BeneficiaryType beneficiaryType;
        try {
            beneficiaryType = BeneficiaryType.valueOf(type.toUpperCase());
        } catch (IllegalArgumentException e) {
            throw new BadRequestException("Type must be 'BANK' or 'USER'", ErrorCode.INVALID_INPUT);
        }
        return ResponseEntity.ok(ApiResponse.success("Beneficiaries retrieved successfully",
                beneficiaryService.getByType(userId, beneficiaryType)));
    }

    @Operation(
        summary     = "Search beneficiaries",
        description = "Full-text search across a user's beneficiaries by name, account number, or username."
    )
    @GetMapping("/{userId}/search")
    @PreAuthorize("hasRole('USER')")
    public ResponseEntity<ApiResponse<List<BeneficiaryDTO>>> search(
            @PathVariable Long userId,
            @RequestParam("q") String q) {
        if (q == null || q.isBlank())
            throw new BadRequestException("Provide a search query parameter", ErrorCode.INVALID_INPUT);
        return ResponseEntity.ok(ApiResponse.success("Search completed successfully",
                beneficiaryService.search(userId, q)));
    }

    @Operation(
        summary     = "Verify a beneficiary",
        description = "Confirms that a saved beneficiary record is still valid and active."
    )
    @GetMapping("/{id}/verify")
    @PreAuthorize("hasRole('USER')")
    public ResponseEntity<ApiResponse<BeneficiaryDTO>> verify(
            @PathVariable Long id,
            @RequestParam Long userId) {
        return ResponseEntity.ok(ApiResponse.success("Beneficiary verified",
                beneficiaryService.getById(id, userId)));
    }

    @Operation(
        summary     = "Get beneficiary by recipient username",
        description = "Looks up a saved USER-type beneficiary by the recipient's username."
    )
    @GetMapping("/{userId}/username/{recipientUsername}")
    @PreAuthorize("hasRole('USER')")
    public ResponseEntity<ApiResponse<BeneficiaryDTO>> getByUsername(
            @PathVariable Long userId,
            @PathVariable String recipientUsername) {
        if (recipientUsername == null || recipientUsername.isBlank())
            throw new BadRequestException("recipientUsername is required", ErrorCode.INVALID_INPUT);
        return ResponseEntity.ok(ApiResponse.success("Beneficiary retrieved successfully",
                beneficiaryService.getByUsername(userId, recipientUsername)));
    }

    @Operation(
        summary     = "Update a beneficiary",
        description = "Replaces an existing beneficiary's details with the supplied data."
    )
    @PutMapping("/{id}")
    @PreAuthorize("hasRole('USER')")
    public ResponseEntity<ApiResponse<BeneficiaryDTO>> update(
            @PathVariable Long id,
            @Valid @RequestBody CreateBeneficiaryRequest req) {
        validateCreateRequest(req);
        return ResponseEntity.ok(ApiResponse.success("Beneficiary updated successfully",
                beneficiaryService.update(id, req)));
    }

    @Operation(
        summary     = "Delete a beneficiary",
        description = "Removes a single beneficiary record for the given user."
    )
    @DeleteMapping("/{id}")
    @PreAuthorize("hasRole('USER')")
    public ResponseEntity<ApiResponse<Void>> delete(
            @PathVariable Long id,
            @RequestParam Long userId) {
        beneficiaryService.delete(id, userId);
        return ResponseEntity.ok(ApiResponse.success("Beneficiary deleted successfully", null));
    }

    @Operation(
        summary     = "Delete multiple beneficiaries",
        description = "Removes a list of beneficiary IDs for the given user in a single request."
    )
    @DeleteMapping("/bulk")
    @PreAuthorize("hasRole('USER')")
    public ResponseEntity<ApiResponse<Void>> deleteByIds(
            @Valid @RequestBody DeleteBeneficiariesRequest req) {
        beneficiaryService.deleteByIds(req.getIds(), req.getUserId());
        return ResponseEntity.ok(ApiResponse.success("Beneficiaries deleted successfully", null));
    }

    private void validateCreateRequest(CreateBeneficiaryRequest req) {
        if (req.getBeneficiaryType() == BeneficiaryType.BANK) {
            if (req.getAccountNumber() == null || req.getAccountNumber().isBlank())
                throw new BadRequestException("accountNumber is required for bank beneficiaries", ErrorCode.INVALID_INPUT);
            if (req.getAccountName() == null || req.getAccountName().isBlank())
                throw new BadRequestException("accountName is required for bank beneficiaries", ErrorCode.INVALID_INPUT);
            if (req.getBankCode() == null || req.getBankCode().isBlank())
                throw new BadRequestException("bankCode is required for bank beneficiaries", ErrorCode.INVALID_INPUT);
            if (req.getBankName() == null || req.getBankName().isBlank())
                throw new BadRequestException("bankName is required for bank beneficiaries", ErrorCode.INVALID_INPUT);
        } else if (req.getBeneficiaryType() == BeneficiaryType.USER) {
            if (req.getRecipientUsername() == null || req.getRecipientUsername().isBlank())
                throw new BadRequestException("recipientUsername is required for user beneficiaries", ErrorCode.INVALID_INPUT);
        }
    }
}
